internal/exec/oci_utils.go

```diff
@@ -1,102 +1,125 @@
 	ref, err := name.ParseReference(imageName)
 	if err != nil {
-		return fmt.Errorf("cannot parse reference of the image '%s'. Error: %v", imageName, err)
+		log.Error("Failed to parse OCI image reference", "image", imageName, "error", err)
+		return fmt.Errorf("invalid image reference: %w", err)
 	}
 
-	// Get the image descriptor
-	descriptor, err := remote.Get(ref)
+	descriptor, err := pullImage(ref)
 	if err != nil {
-		return fmt.Errorf("cannot get image '%s'. Error: %v", imageName, err)
+		return fmt.Errorf("failed to pull image: %w", err)
 	}
 
-	// Download the image from the OCI registry
-	image, err := descriptor.Image()
+	img, err := descriptor.Image()
 	if err != nil {
-		return fmt.Errorf("cannot get a descriptor for the OCI image '%s'. Error: %v", imageName, err)
+		log.Error("Failed to get image descriptor", "image", imageName, "error", err)
+		return fmt.Errorf("cannot get a descriptor for the OCI image '%s': %w", imageName, err)
 	}
 
-	// Write the image tarball to the temp directory
-	err = tarball.WriteToFile(tempTarFileName, ref, image)
+	layers, err := img.Layers()
 	if err != nil {
-		return err
+		log.Error("Failed to retrieve layers from OCI image", "image", imageName, "error", err)
+		return fmt.Errorf("failed to get image layers: %w", err)
 	}
 
-	// Get the tarball manifest
-	m, err := tarball.LoadManifest(func() (io.ReadCloser, error) {
-		f, err := os.Open(tempTarFileName)
+	if len(layers) == 0 {
+		log.Warn("OCI image has no layers", "image", imageName)
+		return fmt.Errorf("the OCI image '%s' does not have any layers", imageName)
+	}
+
+	for i, layer := range layers {
+		// Get the layer's media type
+		layerMediaType, err := layer.MediaType()
 		if err != nil {
-			u.LogError(err)
-			return nil, err
+			log.Warn("Failed to get media type for layer", "index", i, "error", err)
+			continue
+		}
+
+		// Skip layers that don't match the target artifact type
+		if layerMediaType != targetArtifactType {
+			log.Debug("Skipping layer due to media type mismatch", "index", i, "media_type", layerMediaType)
+			continue
+		}
+		if err := processLayer(layer, i, destDir); err != nil {
+			return fmt.Errorf("failed to process layer %d: %w", i, err)
 		}
-		return f, nil
-	})
-	if err != nil {
-		return err
 	}
 
-	if len(m) == 0 {
-		return fmt.Errorf(
-			"the OCI image '%s' does not have a manifest. Refer to https://docs.docker.com/registry/spec/manifest-v2-2",
-			imageName)
+	return nil
+}
+
+// pullImage pulls an OCI image from the specified reference and returns its descriptor.
+func pullImage(ref name.Reference) (*remote.Descriptor, error) {
+	githubToken := os.Getenv(githubTokenEnv)
+	if githubToken == "" {
+		log.Debug("Missing GITHUB_TOKEN environment variable")
 	}
 
-	manifest := m[0]
+	var opts []remote.Option
+	if githubToken != "" {
+		opts = append(opts, remote.WithAuth(&authn.Basic{
+			Username: "oauth2",
+			Password: githubToken,
+		}))
+		log.Debug("Using GitHub token for authentication")
+	}
```

Comment:
this should be wrapped in a conditional and only used if `ghcr.io` is the URI host.

Reply:
We should default to anonymous pulls

```
	opts = append(opts, remote.WithAuth(authn.Anonymous))
```